Modularise and add more notification sources

dislocated-su opened the pull request:
Hi @oelcode 

I rebased the changes from the test branch onto main and merged them with the config updates along with some minor cleanup.

Thought this PR might be a good way to track discussion and work on the refactor as well as enable both of us to push somewhere. Feel free to add to it.

One minor idea I have for now is to use python's logger for console logging (not stock updates). That way the log level can be changed easily, we can auto-format log messages with timestamps, and exceptions will have more log output by default.

Let know what you think!

oelcode committed:
Refactored 50check.py for modular

oelcode committed:
Updated requirements.txt

oelcode committed:
Handler initialiser

oelcode committed:
Modular handlers for notifications

Renamed handler folder to correct name.

oelcode committed:
Fixed ntfy handler.

Fixed ntfy handler so that notification content now appears as intended. Also added support for basic auth, and access token auth.

dislocated-su committed:
Fix discord-webhook handler

dislocated-su committed:
Install optional requirements by default

Could also create an optional-requirements.txt but for now this makes the setup process easier.

dislocated-su commented:
Sorry for the force-pushes on a shared branch, I was working on a different machine and the commit author was unset.

oelcode commented:
Hey @dislocated-su - I have some time to look at this properly this weekend. Are you still working on code at the moment?

dislocated-su commented:
Yeah I've got a few changes I haven't pushed, one is slightly polishing the ntfy handler, another fixes health status updates (they aren't sent at the moment). I'll push when I have the chance tomorrow

I think more could be done: using Python's logger module, maybe per-source status/health updates (like we have now in main). There's also documentation both in config.py and the README.

dislocated-su committed:
fix status messages not sending

probably not the final solution for health updates, as the script previously supported per-source customisation

dislocated-su committed:
add support for mentions in the discord webhooks

dislocated-su committed:
send stock notifications concurrently

dislocated-su commented on handlers/__init__.py:
```diff
@@ -0,0 +1,134 @@
+from abc import ABC, abstractmethod
```
I think this file might be best moved to something like `notification.py` outside of this directory. It does very little common `__init__.py` tasks (initialising the package, exporting methods and classes). Thoughts?

dislocated-su commented on handlers/home_assistant.py:
```diff
@@ -0,0 +1,154 @@
+
+        await self._send_notification(notification_data)
+
+    def format_duration(self, duration):
```
This could go into `notification.py` as I see it is reused across multiple files in `handlers`

dislocated-su committed:
console handler health updates and update to log_stock_checks

dislocated-su committed:
handle shutdown signals without crashing

dislocated-su committed:
update README as ntfy now sends startup and status messages

oelcode commented:
Hey @dislocated-su - still working on this?

dislocated-su commented:
No more changes from my end. I did submit some review comments for you to look at though. Could prompt a bit more work.
